src/script/lua_api/l_object.cpp

```diff
@@ -1806,6 +1806,54 @@ int ObjectRef::l_get_day_night_ratio(lua_State *L)
 	return 1;
 }
 
+int ObjectRef::l_set_camera_mode_restriction_flags(lua_State *L)
+{
+	NO_MAP_LOCK_REQUIRED;
+	ObjectRef *ref = checkobject(L, 1);
+	RemotePlayer *player = get_player(ref);
+	if (!player)
+		return 0;
+
+	// Do it
+	bool first, third, third_front;
+	getboolfield(L, 2, "first", first);
+	getboolfield(L, 2, "third", third);
+	getboolfield(L, 2, "third_front", third_front);
+
+	u16 restrictions = 0;
+	if (first)
+		restrictions |= 0x01;
+	if (third)
+		restrictions |= 0x02;
+	if (third_front)
+		restrictions |= 0x04;
+
+	player->setCameraModeRestrictions(restrictions);
+	getServer(L)->SendCameraModeRestrictions(player, restrictions);
+	return 0;
+}
+
+int ObjectRef::l_get_camera_mode_restriction_flags(lua_State *L)
+{
+	NO_MAP_LOCK_REQUIRED;
+	ObjectRef *ref = checkobject(L, 1);
+	RemotePlayer *player = get_player(ref);
+	if (!player)
+		return 0;
+
+	u16 restrictions = player->getCameraModeRestrictions();
+
+	lua_newtable(L);
+	lua_pushboolean(L, restrictions & 0x01);
```

Comment:
`camera.h` already contains the `CameraModes` enum, and I didn't want to duplicate the exact same enum elsewhere in the code. Should I move the existing enum to a common (i.e. accessible by both server and client) place, say, `player.h`?

Reply:
I'd prefer to see something like this using `string_to_enum` and a `set` of the available camera modes.
```C++
const struct EnumString es_CameraModes[] =
{
	{CAMERA_MODE_FIRST,       "first"},
	{CAMERA_MODE_THIRD,       "third"}
	{CAMERA_MODE_THIRD_FRONT, "front"},
	{0, NULL},
};

std::set<CameraMode> m_camera_modes;

int ObjectRef::l_set_camera_modes(lua_State *L)
{
	// ...

	if (!lua_istable(L, 1))
		return 0; // must be like {}

	lua_pushnil(L);
	// Possibly 1, maybe some other index
	while (lua_next(L, 1) != 0) {
		int enum_val;
		if (string_to_enum(es_CameraModes, enum_val, lua_tostring(L, -2))) {
			if (readParam<bool>(L, -1, true))
				m_camera_modes.emplace((CameraMode)enum_val);
			else
				m_camera_modes.erase((CameraMode)enum_val);
		}
		lua_pop(L, 1);
	}

	// Needs at least one
	if (m_camera_modes.count() == 0)
		m_camera_modes.emplace(CAMERA_MODE_FIRST);
}

void Camera::toggleCameraMode()
{
	auto it = m_camera_modes.find(m_camera_mode);
	if (it == m_camera_modes.end()) {
		m_camera_mode = *m_camera_modes.begin();
		return;
	}

	if (++it == m_camera_modes.end())
		it = m_camera_modes.begin();
	m_camera_mode = *it;
}
```

Reply:
Setting the camera modes themselves is much better than setting "restrictions" for the camera modes. :+1: 

Will do.

Reply:
Please also suggest the best place to declare `std::set<CameraMode> m_camera_modes`. ;)

From your code above, I think it could be declared in `l_object.cpp`, but the problem is that `Camera::toggleCameraMode` won't be able to access it directly, as above.

How about in the `Player` class? Both the server and the client would store the camera modes in `RemotePlayer` and `LocalPlayer` respectively.

Reply:
`Player` is probably the best choice. I'd use a getter function (non-const reference to the set) in this case to expose it to the Lua part.

Reply:
Sounds good.

Reply:
@SmallJoker How would these be sent over the network? Currently, these modes are encoded and sent as a `u16` over to the client, where the value is decoded again.

Reply:
As I'm not familiar with `EnumString`, would you also be able to help me understand how the getter API method would work?

Reply:
@ClobberXD You can either continue to send the entire vector (starting with length=`u16`) as `u16` (enum values) or `std::string` (Lua names). Both options would be OK.
Regarding the EnumString functions: 
 * Get enum pendant for a string value: `string_to_enum(enumstring_array, out_as_int, input_string)`
 * Get string pendant for an enum value: [TODO] `enum_to_string(enumstring_array, out_as_string, input_enum_as_int)` 

The idea of the code above would be that modders can indicate the changes of the available camera modes. If one mode is not specified in the setter function, it won't be updated.
The getter function would then push a table containing all camera modes: (iterate though all available modes until CAMERA_MODE_MAX (TODO).
```Lua
{
	first = true, -- contained in m_camera_modes
	third = false, -- not in m_camera_modes
	front = true
}
```

I see this is getting quite a long story for such a small feature. I'm sorry if this is tiring.
FYI: `luamethod(ObjectRef, set_camera_mode_restriction_flags),` needs updating in your source as well.

Reply:
> You can either continue to send the entire vector (starting with length=u16) as u16 (enum values) or std::string (Lua names).

Alright, will look into it.

> Get string pendant for an enum value: **[TODO]** 
> `enum_to_string(enumstring_array, out_as_string, input_enum_as_int)`

Seeing `string_to_enum`, I already tried looking for `enum_to_string`, but couldn't find it.

> I see this is getting quite a long story for such a small feature. I'm sorry if this is tiring.

This is indeed getting too draggy, but I wholeheartedly support your approach as it's much more straightforward. The end result would be very robust and easy to work with, and that makes the wasted efforts worthwhile. Thanks for all your help. :)

>  luamethod(ObjectRef, set_camera_mode_restriction_flags), needs updating in your source as well.

Will do, thanks.

Reply:
> The getter function would then push a table containing all camera modes: (iterate though all available modes until CAMERA_MODE_MAX (TODO)

I was wondering if there's a better way to iterate through all available modes. Maybe convert the set into a map (`std::map`)? Each player will own a copy of this map, and each key can be assigned `true` or `false`.